### ISSUE TITLE
Single layer  dv

### ISSUE THREAD

**Issue by roboluzzo** [DolbyLaboratories/dlb_mp4base#15]
Hi, is possible to make single layer from 2 separated track?
Thanks, Roberto.

**Comment by ShaoWeiguo** [comment]
Just clarify,  do you want to mux the mp4 into single track from 2 separated ES ? Or generate the ES into single layer from 2 separate ES?

**Comment by MastaG** [comment]
I think he's referring to the second.. generate a single ES from two separate ES's (most likely the bluray base-layer and DV enhanced layer).
There's already a public tool available for that written in python by yusesope.
https://www.makemkv.com/forum/viewtopic.php?f=12&t=18602&hilit=atmos&start=1335#p85413

Then you can mux the resulting file with mp4muxer using --dv-profile 7 (still have to try that).

Still it would be nice if this functionality could be added to mp4muxer.

**Comment by roboluzzo** [comment]
Thanks for your attention, I asked you about the single layer because the various media players like nvidia shield, work only with that type of file. The dual layers only work from USB or NAS connected to the TV and the readers that can be installed on TVs like mine are not very comfortable to use.
Thank you, Roberto.
________________________________
Da: MastaG <notifications@github.com>
Inviato: martedì 14 aprile 2020 08:30
A: DolbyLaboratories/dlb_mp4base <dlb_mp4base@noreply.github.com>
Cc: roboluzzo <roboluzzo@hotmail.it>; Author <author@noreply.github.com>
Oggetto: Re: [DolbyLaboratories/dlb_mp4base] Single layer dv (#15)


I think he's referring to the second.. generate a single ES from two separate ES's (most likely the bluray base-layer and DV enhanced layer).
There's already a public tool available for that written in python by yusesope.
https://www.makemkv.com/forum/viewtopic.php?f=12&t=18602&hilit=atmos&start=1335#p85413

Then you can mux the resulting file with mp4muxer using --dv-profile 7 (still have to try that).

Still it would be nice if this functionality could be added to mp4muxer.

**Comment by RonRicoSuave** [comment]
As far as I can tell from the "yusescope" tool's source, all it's doing when "single layer" mode is selected is discarding the enhancement layer entirely.  Since profile 7 requires an enhancement layer of some type (even though with MEL it's just blank), all this is doing is creating a broken profile 7 file with no EL that happens to work on devices that aren't doing a very good job of checking stream validity or were ignoring the EL anyway.   Forcing it to profile 8.1 at least produces something that conforms to the profile it's in.

I'd be more interested in Dolby's muxer being able to accept a base layer and enhancement layer as input, although I can't think of a great use-case that wouldn't be better solved by grabbing one of the newer boxes that can fully decode BL + FEL images.

Creating a "real" single layer file out of a BL+FEL source would involve doing the appropriate math on the layers and re-encoding and since nothing will play 12bpc HEVC properly (much less come anywhere close to displaying 12 bit color) and the chroma gains are more or less non-existent (what are we at since both layers are 4:2:0, about 1/16 "true" 12bpc chroma res and 1/4 luma?)...   there's not much real point in going that route.

**Comment by RonRicoSuave** [comment]
Oh, I see what you mean, I'd entirely forgotten on the structure of some disks the layers are separate tracks, the mode of the tool I'm using pulls either format as a single layer HEVC stream in MKV.   I must have been looking at an old version(?) of the yusescope source code, I think, because it definitely wasn't doing what you're talking about in "single layer" mode.  It just skipped all the EL NALs.   

The LG Nano will go into DV mode with AAC audio (have you tried it on yours?) and I suspect it would work with MP4 + opus if anything that didn't break DV could mux it in.  Opus is exactly nowhere in the supported codecs list and works in every container type it's compatible with that I've tried it in, at least with HDR10.  Certain audio types will cause DV streams to fail to even trigger HDR10 on those TVs so they're kinda weird.  I normally do what you do but with dovi_tool and AAC audio since it's supposed to be better quality than AC3 and AAC in .m2ts + DV HEVC will trigger DolbyVision mode properly on the TV.  I gave up on keeping around the HD audio and just encode the AAC at extremely high bitrates until I give up and spring for one of the new streaming boxes that can decode FEL.  I'm only avoiding a physical UHD Bluray player because I've already seen a barely noticeable spec of dust cause tons of seeking and re-read attempts on a brand new disk's first time in the drive...  I really don't want to be changing those out all the time....    plus a lot of these are getting Rifftrax added in to the audio as the primary track -- I'm buying lots of terrible movies thanks to those guys.  :P



From what I read the C8s were the last LG OLEDs that supported playing profile 7 anyway.

JW3 has a really high bitrate on top of the thing where LG TVs max allowable bitrate drops by 10Mb/s for dolby vision files so it doesn't surprise me that it would give you problems -- It's roughly 71Mb/s video-only with EL and a new enough title that it might be using more complex DV metadata, I didn't examine it.   LG TVs have some fairly low limits for allowable bitrate when streaming over the network that goes down to ~50Mb/s when it's a DV stream for whatever reason, at least according to their manuals.  

Technically the output of all Dolby Vision is 12bpc with or without the EL because of upconversion to HDR colorspace when the transforms are applied (check the output depth of a profile 7 BL+MEL and it's still 12bpc even though the MEL is an empty video), and like I said above the EL only gets you an extra 2bpc of real color info per 4x4 pixels (applied to the already 1/2 res chroma of 4:2:0 video) and an extra 2bpc of real luma per 2x2 block.  The trick is that 4k resolution is so high from any reasonable viewing distance that you won't notice, which is part of how so many UHD BDs are getting away with releasing as upscaled 2k masters without people complaining.  Actual 12bpc HEVC stutters like crazy on my PC and I can do realtime playback of lossless 12bpc JPEG2000 video (which is enormous and non-trivial to decode) so I have no clue how terrible the 12 bit software decoder for HEVC must be, so we're a ways off from that until someone starts mass-producing 12bpc decoder chips and by that point they'll probably have moved on to H.266 or something even more painfully slow to encode.
